platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/internal/reconnect/ConcurrentBlockingIterator.java

```diff
@@ -109,17 +109,15 @@ public boolean hasNext() {
         // if closed and buffer.poll != null || !closed
         final long waitMillis = maxWaitTimeUnit.toMillis(maxWaitTime);
         final long timeOutWhenMillisAre = System.currentTimeMillis() + waitMillis;
-        while ((next = buffer.poll()) == null) {
-            if (closed.get()) {
-                return false;
-            } else {
-                if (System.currentTimeMillis() > timeOutWhenMillisAre) {
-                    throw new RuntimeException(new TimeoutException("Timed out trying to read from buffer"));
-                }
+        boolean isOpen = !closed.get();
+        while (((next = buffer.poll()) == null) && isOpen) {
```

Comment:
Given the time resolution for the wait is milliseconds, it _might_ be worth considering using the timeout version of [poll](https://docs.oracle.com/en/java/javase/21/docs/api/java.base/java/util/concurrent/LinkedBlockingQueue.html#poll(long,java.util.concurrent.TimeUnit)) with a 1 millisecond timeout.  That would still check the closed value every millisecond, but avoid the thousands of additional loops in between.

Also, could this be rewritten as follows (I think it makes sense, but you might have a reason for splitting it out)?
```
        while (((next = buffer.poll()) == null) && !closed.get()) {
```

Reply:
There is a reason. If written like above, thread context switch may happen after the call to `poll()`, but before the call to `closed.get()`. This is exactly the problem addressed in this PR. With the way how it's currently in the PR, a call to `poll()` is guaranteed after `closed.get()` returns true.

Reply:
Regarding the `poll()` calls, let me file a low-priority ticket about it. As I wrote above in response to @anthony-swirldslabs comments, the chances are we don't need a busy loop here at all

Reply:
Filed https://github.com/hashgraph/hedera-services/issues/11691